oci/oci_dest.go

```diff
@@ -0,0 +1,168 @@
+package oci
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/projectatomic/skopeo/manifest"
+	"github.com/projectatomic/skopeo/reference"
+	"github.com/projectatomic/skopeo/types"
+)
+
+type ociManifest struct {
+	SchemaVersion int               `json:"schemaVersion"`
+	MediaType     string            `json:"mediaType"`
+	Config        descriptor        `json:"config"`
+	Layers        []descriptor      `json:"layers"`
+	Annotations   map[string]string `json:"annotations"`
+}
+
+type descriptor struct {
+	Digest    string `json:"digest"`
+	MediaType string `json:"mediaType"`
+	Size      int64  `json:"size"`
+}
+
+type ociImageDestination struct {
+	dir string
+	ref reference.Named
+	tag string
+}
+
+var refRegexp = regexp.MustCompile(`^([A-Za-z0-9._-]+)+$`)
+
+// NewOCIImageDestination returns an ImageDestination for writing to an existing directory.
+func NewOCIImageDestination(dir string) (types.ImageDestination, error) {
+	dest := dir
+	sep := strings.LastIndex(dir, ":")
+	tag := "latest"
+	if sep != -1 {
+		dest = dir[:sep]
+		tag = dir[sep+1:]
+		if !refRegexp.MatchString(tag) {
+			return nil, fmt.Errorf("Invalid reference %s", tag)
+		}
+	}
+	return &ociImageDestination{
+		dir: dest,
+		tag: tag,
+	}, nil
+}
+
+func (d *ociImageDestination) CanonicalDockerReference() (string, error) {
+	return "", fmt.Errorf("Can not determine canonical Docker reference for an OCI image")
+}
+
+func createManifest(m []byte) ([]byte, string, error) {
+	om := ociManifest{}
+	mt := manifest.GuessMIMEType(m)
+	switch mt {
+	case manifest.DockerV2Schema1MIMEType:
+		return nil, "", fmt.Errorf("not yet implemented")
```

Comment:
I think I'll leave _yet_ because in theory it should be possible to convert a v2s1 to v2s2 as it's what Docker I believe does when pulling busybox for instance (busybox still serve v2s1 form the Docker Hub). Makes sense?